torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -374,6 +374,361 @@ def collect_bw_donated_buffer_idxs(
+    #   outputs - (*grad_outs, *fw_outs)
+    # New signature
+    #   inputs - (*saved_tensors, *tangents) -- Different
+    #   outputs - (*grad_outs)  -- Different
+    # Here both input and output signature change. The output signature handling
+    # is quite easy because the grads_out are sitting at the right place, so we
+    # dont have to do anything.
+    #
+    # For the input signature, we have to collect the saved tensors from the
+    # corresponding forward graph output. We collect all saved_tensors when we
+    # see the forward graph, and save it into a map and then later use it during
+    # the backward.
+
+    env = {}  # type: ignore[var-annotated]
+    new_graph = torch.fx.Graph()
+
+    already_added_new_hop_mods = set()
+
+    # The stack of fw_nodes for invoke_subgraph HOP. There is an implicit
+    # assumption about the graph structure, i.e., if we have hop1, hop2, hop3,
+    # ... in the forward part of the joint graph, we will have .., hop3, hop2,
+    # hop1 order for the backward. This structure allows us to just use a stack
+    # to collect all the information that we need to pass from the forward hop
+    # node to the corresponding backward node.
+    old_fw_hop_nodes_stack = []
+    # Collect the saved tensor nodes that we need to pass as inputs to the
+    # backward hop.
+    fw_node_to_saved_tensors_map = {}
+
+    def add_new_hop_gm(new_subgraph_mod, name):
+        new_subgraph_attr_name = f"{name}_post_graph"
+        if new_subgraph_attr_name in already_added_new_hop_mods:
+            return new_subgraph_attr_name
+
+        joint_gm.register_module(new_subgraph_attr_name, new_subgraph_mod)
+        already_added_new_hop_mods.add(new_subgraph_attr_name)
+        return new_subgraph_attr_name
+
+    def propagate_meta_info(new_hop_gm, new_call_function_node, old_call_function_node):
+        # Copy all the fields from the old call_function node. And then override
+        # the `val` meta field with the outputs of new_hop_gm.
+        new_call_function_node.meta = copy.copy(old_call_function_node.meta)
+
+        output = new_hop_gm.graph.find_nodes(op="output")[0]
+        out_example_vals = [n.meta["val"] if n else None for n in output.args[0]]
+        new_call_function_node.meta["val"] = tuple(out_example_vals)
+
+    for node in joint_gm.graph.nodes:
+        if (
+            node.op == "call_function"
+            and node.target is invoke_subgraph
+            and isinstance(node.args[1], str)
+        ):
+            identifier = (
+                node.args[1].replace("___forward", "").replace("___backward", "")
+            )
+            if node.args[1].startswith("___forward"):
+                # Insert the new_fw_hop_gm. This is straightforward. Get the
+                # new_fw_hop_gm, insert the hop_gm as a get_attr node, and then
+                # add a call_function node. Additionally, also use getitem
+                # call_functions to collect the saved_tensor nodes
+
```

Comment:
I think this should be possible to do in-place. We can add the new node, replace all usges of the new_fwd_outs with the old_fwd_outs, and we're done.